### ISSUE TITLE
NRT thread can block disk I/O thread

### ISSUE THREAD

**Issue by Spacechild1** [supercollider/supercollider#5511]
The disk I/O thread in `DiskIO_UGens.cpp` locks the NRT thread when it wants to write data from the buffer to the disk. AFAICT, this is only necessary because the user might accidentally free the buffer concurrently (which shouldn't really happen).

I understand that (ab)using Server sound buffers for disk I/O is convenient because it already handles the opening and closing of soundfiles. However, it also means the disk I/O thread is implicitly tied to the NRT thread, which kind of defeats the whole purpose of having a dedicated disk I/O thread in the first place...

Note that asynchronous commands executed on the NRT thread can take an arbitrary amount of time!

This causes problems such as this: https://git.iem.at/pd/vstplugin/-/issues/58

**Comment by joshpar** [comment]
This lock was added, I think, to make sure the disk io ugens worked in NRT mode.

/*
Josh Parmenter
www.realizedsound.net/josh
*/

**Comment by Spacechild1** [comment]
No, `DiskIn` and `VDiskIn` contain different code paths for RT and NRT synthesis. The latter doesn't use the disk I/O thread at all.

---

BTW, I find it strange that `DiskOut` *does* use the disk I/O thread in NRT mode. There is basically no synchronization between the main thread and the I/O thread, so if the latter can't catch up with the former, it would drop audio... I think it should write to disk synchronously instead.

**Comment by joshpar** [comment]
I'm surprised. I've used DiskOut a LOT for large NRT files without any problems. Regardless - if you propose any changes, please make sure reading and writing for those UGens in NRT still works. Thanks!

**Comment by Spacechild1** [comment]
> I've used DiskOut a LOT for large NRT files without any problems

That's interesting to know! In practice it's not a problem if the buffer is sufficiently big, but it's definitely not safe. 

> please make sure reading and writing for those UGens in NRT still works

Sure, good point!
